x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/AnalyticsProcessManager.java

```diff
@@ -159,12 +162,75 @@ private void refreshDest(DataFrameAnalyticsConfig config) {
             () -> client.execute(RefreshAction.INSTANCE, new RefreshRequest(config.getDest().getIndex())).actionGet());
     }
 
-    static class ProcessContext {
+    public void stop(TransportStartDataFrameAnalyticsAction.DataFrameAnalyticsTask task) {
+        ProcessContext processContext = processContextByAllocation.get(task.getAllocationId());
+        if (processContext != null) {
+            LOGGER.debug("[{}] Stopping process", task.getParams().getId() );
+            processContext.stop();
+        } else {
+            LOGGER.debug("[{}] No process context to stop", task.getParams().getId() );
+        }
+    }
+
+    class ProcessContext {
 
+        private final String id;
+        private volatile AnalyticsProcess process;
+        private volatile DataFrameDataExtractor dataExtractor;
+        private volatile AnalyticsResultProcessor resultProcessor;
         private final AtomicInteger progressPercent = new AtomicInteger(0);
+        private volatile boolean processKilled;
+
+        ProcessContext(String id) {
+            this.id = Objects.requireNonNull(id);
+        }
+
+        public String getId() {
+            return id;
+        }
+
+        public boolean isProcessKilled() {
+            return processKilled;
+        }
 
         void setProgressPercent(int progressPercent) {
             this.progressPercent.set(progressPercent);
         }
+
+        public synchronized void stop() {
+            LOGGER.debug("[{}] Stopping process", id);
+            processKilled = true;
+            dataExtractor.cancel();
```

Comment:
this should check for `dataExtractor != null`.

Reply:
Good point. Same is true for the `process` below. I'll add null checks.